### PR TITLE
Scheduler: reduce iteration and conversion to improve perfermance

### PR DESCRIPTION
**Pull request by shiyan2016**
rewrite two funcs:
1. GetNamespacesFromPodAffinityTerm func
2. PodMatchesTermsNamespaceAndSelector func

**What type of PR is this?**
/kind design
/sig scheduling

**What this PR does / why we need it**:
This PR tries to reduce iteration and remove affinity term namespaces  convert to map and labelSelector convert to selector, just use term namespaces and labelSelector to match pod namespace and labels. Which can improve the scheduler perfermance:

- Required PodAntiAffinity (2.4X perfermance improvement, as below)
  - Before
  ```
  BenchmarkSchedulingPodAntiAffinity/1000Nodes/1000Pods-4       10000	  24375155 ns/op
  ```
  - After
  ```
  BenchmarkSchedulingPodAntiAffinity/1000Nodes/1000Pods-4       10000	   9812070 ns/op
  ```
**Special notes for your reviewer**:
Above test result is run at a bm with 4 core cpus and 8GB memory.

**Does this PR introduce a user-facing change?**:
2.4X perfermance improvement on PodAntiAffinity

### CODE REVIEW

**Comment by wgliang** [comment]
/ok-to-test
@shiyan2016 Thanks, I will review it.

File: pkg/scheduler/algorithm/predicates/metadata.go [hunk 1]
```diff
@@ -396,6 +392,7 @@ func getTPMapMatchingExistingAntiAffinity(pod *v1.Pod, nodeInfoMap map[string]*s
 		node := nodeInfo.Node()
 		if node == nil {
 			catchError(fmt.Errorf("node not found"))
+			cancel()
```
Comment (wgliang) [inline pkg/scheduler/algorithm/predicates/metadata.go]: The cancel() operation does not seem to be related to the PR.

Comment (wgliang) [reply pkg/scheduler/algorithm/predicates/metadata.go]: If this is a issue, I suggest a separate PR to complete in order to better track.

File: pkg/scheduler/algorithm/priorities/util/topologies.go [hunk 2]
```diff
@@ -18,36 +18,91 @@ package util
...
 
+func matchLabels(ls labels.Labels, key string, value string) bool {
+	if !ls.Has(key) {
+		return false
+	}
+	return value == ls.Get(key)
+}
+
+func expressionHasValue(expr metav1.LabelSelectorRequirement, value string) bool {
+	for i := range expr.Values {
```
Comment (wgliang) [inline pkg/scheduler/algorithm/priorities/util/topologies.go]: why not just do it like:
`for i,v := range expr.Values`

File: pkg/scheduler/algorithm/predicates/metadata.go (hunk 1 above)
Comment (shiyan2016) [reply pkg/scheduler/algorithm/predicates/metadata.go]: ok

Comment (shiyan2016) [reply pkg/scheduler/algorithm/predicates/metadata.go]: > The cancel() operation does not seem to be related to the PR.

done

File: pkg/scheduler/algorithm/priorities/util/topologies.go (hunk 2 above)
Comment (shiyan2016) [reply pkg/scheduler/algorithm/priorities/util/topologies.go]: done

File: pkg/scheduler/algorithm/priorities/util/topologies.go [hunk 3]
```diff
@@ -18,36 +18,91 @@ package util
...
-	return names
+	return []string{pod.Namespace}
 }
 
 // PodMatchesTermsNamespaceAndSelector returns true if the given <pod>
 // matches the namespace and selector defined by <affinityPod>`s <term>.
-func PodMatchesTermsNamespaceAndSelector(pod *v1.Pod, namespaces sets.String, selector labels.Selector) bool {
-	if !namespaces.Has(pod.Namespace) {
-		return false
+func PodMatchesTermsNamespaceAndSelector(pod *v1.Pod, namespaces []string, ps *metav1.LabelSelector) bool {
```
Comment (congliu01) [inline pkg/scheduler/algorithm/priorities/util/topologies.go]: `ps` is a bit confusing, consider keeping `selector`?

File: pkg/scheduler/algorithm/priorities/util/topologies.go [hunk 4]
```diff
@@ -18,36 +18,91 @@ package util
...
+	for _, ns := range namespaces {
+		if pod.Namespace == ns {
+			if LabelsMatchLabelSelector(labels.Set(pod.Labels), ps) {
+				return true
+			}
```
Comment (congliu01) [inline pkg/scheduler/algorithm/priorities/util/topologies.go]: I think you can add `else {return false}` here, otherwise we need to iterate over the remaining namespaces which is useless.

File: pkg/scheduler/algorithm/priorities/util/topologies.go [hunk 5]
```diff
@@ -18,36 +18,91 @@ package util
...
+		}
 	}
 
-	if !selector.Matches(labels.Set(pod.Labels)) {
+	return false
+}
+
+// LabelsMatchLabelSelector returns true if the given <labels> matches the labelSelector
+func LabelsMatchLabelSelector(ls labels.Labels, ps *metav1.LabelSelector) bool {
```
Comment (congliu01) [inline pkg/scheduler/algorithm/priorities/util/topologies.go]: `ls` -> `labels`, `ps` -> `labelSelector`? this matches the function comment too.

**Review by liu-cong — COMMENTED** [review]
It seems there are 2 major changes: 1. avoid converting namespace string list to string set and 2. avoid converting LabelSelector to a Selector. Just out of curiosity, have you characterized the performance improvement of each change? which one is dominating?

**Comment by Huang-Wei** [comment]
It seems this PR doesn't really improve perf:

## Without this PR

```
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000      98,540,988 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000     102,982,389 ns/op

BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      34,792,657 ns/op
BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      34,314,487 ns/op
```

## With this PR

```
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000     102,279,957 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000     104,050,594 ns/op

BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      35,028,057 ns/op
BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      34,455,011 ns/op
```

**Comment by shiyan2016** [comment]
> It seems this PR doesn't really improve perf:
> 
> ## Without this PR
> ```
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000      98,540,988 ns/op
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000     102,982,389 ns/op
> 
> BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      34,792,657 ns/op
> BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      34,314,487 ns/op
> ```
> 
> ## With this PR
> ```
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000     102,279,957 ns/op
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000     104,050,594 ns/op
> 
> BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      35,028,057 ns/op
> BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      34,455,011 ns/op
> ```



> It seems this PR doesn't really improve perf:
> 
> ## Without this PR
> ```
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000      98,540,988 ns/op
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000     102,982,389 ns/op
> 
> BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      34,792,657 ns/op
> BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      34,314,487 ns/op
> ```
> 
> ## With this PR
> ```
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000     102,279,957 ns/op
> BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8     1000     104,050,594 ns/op
> 
> BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      35,028,057 ns/op
> BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8     1000      34,455,011 ns/op
> ```

When schedule 10000 pods in a 1000 node cluster, it can improve perfermance (like PR https://github.com/kubernetes/kubernetes/pull/67788, https://github.com/kubernetes/kubernetes/pull/74041 test). See the above test result:
```
  - Before
    BenchmarkSchedulingPodAntiAffinity/1000Nodes/1000Pods-4       10000	  24375155 ns/op
  - After
   BenchmarkSchedulingPodAntiAffinity/1000Nodes/1000Pods-4        10000	   9812070 ns/op
```

File: pkg/scheduler/algorithm/priorities/util/topologies.go (hunk 3 above)
Comment (shiyan2016) [reply pkg/scheduler/algorithm/priorities/util/topologies.go]: done

File: pkg/scheduler/algorithm/priorities/util/topologies.go (hunk 4 above)
Comment (shiyan2016) [reply pkg/scheduler/algorithm/priorities/util/topologies.go]: done

File: pkg/scheduler/algorithm/priorities/util/topologies.go (hunk 5 above)
Comment (shiyan2016) [reply pkg/scheduler/algorithm/priorities/util/topologies.go]: done

**Comment by tedyu** [comment]
@shiyan2016 
The performance numbers you posted were before the rebase, right ?

Do you mind running the benchmark on latest master and provide the comparison ?

Thanks

**Comment by shiyan2016** [comment]
> @shiyan2016
> The performance numbers you posted were before the rebase, right ?
> 
> Do you mind running the benchmark on latest master and provide the comparison ?
> 
> Thanks

ok

File: pkg/scheduler/algorithm/predicates/metadata.go (hunk 1 above)
Comment (shiyan2016) [reply pkg/scheduler/algorithm/predicates/metadata.go]: > If this is a issue, I suggest a separate PR to complete in order to better track.

I had opened a new issue for this, ref: https://github.com/kubernetes/kubernetes/pull/79774.

**Comment by shiyan2016** [comment]
> @shiyan2016
> The performance numbers you posted were before the rebase, right ?
> 
> Do you mind running the benchmark on latest master and provide the comparison ?
> 
> Thanks

After rebase, the result are as follows:
```
  - Before
    BenchmarkSchedulingPodAntiAffinity/1000Nodes/1000Pods-4         10000	  25698313 ns/op
  - After
    BenchmarkSchedulingPodAntiAffinity/1000Nodes/1000Pods-4        10000	  10649081 ns/op
```

**Comment by shiyan2016** [comment]
@wgliang @congliu01 Pls help me to review this PR continuely. Thanks very much.

**Review by wgliang — APPROVED** [review]
/lgtm 

Waitting for the CI to be happy

**Comment by shiyan2016** [comment]
@wgliang The  commits squashed and the CI test is ok. Pls have a look. Thanks.

**Comment by wgliang** [comment]
/lgtm
@shiyan2016  Thanks!

**Comment by shiyan2016** [comment]
@wgliang @Huang-Wei  I have done the rebase. Pls have a look.Thanks very much.

File: pkg/scheduler/algorithm/priorities/util/topologies.go (hunk 4 above)
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/priorities/util/topologies.go]: Replace with the whole if...else... block with:

```
return LabelsMatchLabelSelector(labels.Set(pod.Labels), selector)
```

**Comment by Huang-Wei** [comment]
@shiyan2016 I'm having a quite limited time reviewing PRs this week. Sry for that.

After a quick look, my biggest concerns are:

1. Performance. I'm still not 100% sure we can get enough perf gain.
1. This PR changed an interface to a concrete fixed type. Tightening the arguments to a fixed type doesn't sound like a good idea, esp. in terms of testability and flexibility for possible future changes.
1. Added helper code is not equipped with enough unit tests.
1. Those helper code may be moved into apimachenary packages for reusability.

I will do a close look at perf next week and get back.

File: pkg/scheduler/algorithm/priorities/util/topologies.go (hunk 4 above)
Comment (shiyan2016) [reply pkg/scheduler/algorithm/priorities/util/topologies.go]: The else cannot replaced, which is for pod.Namespace != ns. I have replaced
```
if LabelsMatchLabelSelector(labels.Set(pod.Labels), selector) {
				return true
			}
```
 to 			return LabelsMatchLabelSelector(labels.Set(pod.Labels), selector)

**Comment by shiyan2016** [comment]
> @shiyan2016 I'm having a quite limited time reviewing PRs this week. Sry for that.
> 
> After a quick look, my biggest concerns are:
> 
> 1. Performance. I'm still not 100% sure we can get enough perf gain.
> 2. This PR changed an interface to a concrete fixed type. Tightening the arguments to a fixed type doesn't sound like a good idea, esp. in terms of testability and flexibility for possible future changes.
> 3. Added helper code is not equipped with enough unit tests.
> 4. Those helper code may be moved into apimachenary packages for reusability.
> 
> I will do a close look at perf next week and get back.

@Huang-Wei How about the 1,2 concerns now?
